File: hostalive.c
#include <stdlib.h>
#include <stdio.h>
#include <netinet/in.h>
#include <fcntl.h>
#include <netdb.h>
#include <sys/socket.h>
#include <unistd.h>



int main (int argc, const char *argv[]) {
  int socketid;
	struct sockaddr_in server_address;
	struct hostent* hostinfo;
	const char *hostname;
	int port_number;
	int tries=10;			// number of tries to connect
	int timeout=5;			// connecting timeout in seconds
	int sleep_seconds=30;	// sleep before retry
	fd_set fdset;			// file descritor set for monitoring the socket
    struct timeval tv;		// time out struct

	
//TODO: validate arguments
//TODO: get executable name for usage information
//TODO: configurable timeout
//TODO: configurable tries
//TODO: configurable sleep

	if (argc != 3){
		fprintf(stderr,"Usage: <hostname> <port>\n");
		return 1;
	}else {
		hostname = argv[1];
		port_number = strtol(argv[2],(char **)NULL,10);
	}

	
	
	server_address.sin_family = AF_INET;
	
	printf("Resolving %s ...\n",hostname);
	hostinfo = gethostbyname (hostname);
	if (hostinfo == NULL){
		fprintf(stderr,"Could not resolve IP address for %s\n",hostname);
		return 1;
	}else {
		server_address.sin_addr = *((struct in_addr *) hostinfo->h_addr);	 
	}
	
	server_address.sin_port = htons(port_number);
	
	while(tries--){
		socketid = socket(PF_INET,SOCK_STREAM,0);
		
		// setting socket to non-blocking so I can control the timeout
		fcntl(socketid, F_SETFL, O_NONBLOCK);

		if (socketid != -1){
			printf("Trying to connect...\n");
			connect(socketid,(struct sockaddr *)&server_address,sizeof(struct sockaddr_in));
		
			FD_ZERO(&fdset);			// clear set
			FD_SET(socketid, &fdset);	// assign set to socket
			tv.tv_sec = timeout;		// prepare timeout struct for select
			tv.tv_usec = 0;
			
			// monitoring socket state with select usint the file descriptor set and timeout
			if (select(socketid + 1, NULL, &fdset, NULL, &tv) == 1)
			{
				// something happend to socket
				int socket_error;
				socklen_t len = sizeof socket_error;
				
				// getting socket state
				getsockopt(socketid, SOL_SOCKET, SO_ERROR, &socket_error, &len);
			
				if (socket_error == 0) {
					printf("Connected.\n");
					close(socketid);
					//TODO: run something
					return 0;				
				}else {
					perror("Connect");
				}
			}
			else {
				printf("Timeout... Sleeping %d seconds... \n",sleep_seconds);
				sleep(sleep_seconds);
			}

		
		}else {
			fprintf(stderr,"Could not create socket...\n");
			return -1;
		}
	}

    return 0;
}

File: hostalive.py
#!/usr/bin/python

import socket
import sys
import time
import argparse

RETRIES=100
TIMEOUT_SECONDS=10
SLEEP_SECONDS=10

parser = argparse.ArgumentParser(description='Check if host/service is alive.')
parser.add_argument('hostname',metavar='hostname',type=str,nargs=1,help='name of server (short or FQDN)')
parser.add_argument('port',metavar='portnumber',type=int,nargs=1,help='port number for TCP connection')
parser.add_argument('--retries',metavar='retries',type=int,nargs=1,default=[RETRIES],help='number of retries (default is %d)' % RETRIES)
parser.add_argument('--timeout',metavar='timeout',type=int,nargs=1,default=[TIMEOUT_SECONDS],help='seconds to wait for TCP connection (default is %d). Along with --sleep, keep this to a sane high value.' % TIMEOUT_SECONDS)
parser.add_argument('--sleep',metavar='sleep',type=int,nargs=1,default=[SLEEP_SECONDS],help='seconds to wait before retrying TCP connection (default is %d). Along with --timeout, keep this to a sane high value.' % SLEEP_SECONDS)
parser.add_argument('--keepalive',dest='keepalive',action='store_true',help='will continue trying connections even after a successful connection.')
parser.set_defaults(keepalive=False)
args = parser.parse_args()

hostname = args.hostname[0]
port_number = args.port[0]
retries = args.retries[0]
timeout_seconds = args.timeout[0]
sleep_seconds = args.sleep[0]
keepalive = args.keepalive

try:
    server_address=socket.gethostbyname(hostname)
except socket.error as ex:
    sys.stderr.write("Could not resolve hostname!\n")
    quit(1)


socketid = None


while(retries):
    print("Trying to connect to %s - %s on port %d ..." % (hostname,server_address,port_number))
    retries=retries-1
    try:
        socketid = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        socketid.settimeout(timeout_seconds)
    except socket.error as ex:
        sys.stderr.write( "Could not create socket!\n")
        sys.stderr.write(ex.message)
        quit(1)

    try:
        socketid.connect((server_address,port_number))
        print("Connected!")
        socketid.close()
    except socket.error as ex:
        print("Connect failed")
        socketid = None
    if (keepalive and retries) or (socketid == None and retries):
        print("Sleeping %d seconds before retrying..." % sleep_seconds)
        time.sleep(sleep_seconds)
        continue
    socketid = None
    quit(0)

quit(1)
